File: backend/app/schemas/event.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

from app.schemas.chapter import EventChapterResponse
from app.schemas.photo import PhotoVisionResult
from app.schemas.photo_group import PhotoGroupResponse

EventStatus = Literal[
    "clustered", "waiting_for_vision", "ai_pending", "ai_processing", "generated", "ai_failed"
]
StoryFreshness = Literal["fresh", "stale"]
EnhancementStatus = Literal["none", "retained", "expired"]
EventVisionStatus = Literal["pending", "processing", "partial", "completed", "failed", "unsupported"]


class EventVisionSummary(BaseModel):
    status: EventVisionStatus = "pending"
    total: int = 0
    pending: int = 0
    processing: int = 0
    completed: int = 0
    failed: int = 0
    unsupported: int = 0


class EventResponse(BaseModel):
    id: str
    title: str = ""
    locationName: Optional[str] = None
    gpsLat: Optional[float] = None
    gpsLon: Optional[float] = None
    startTime: Optional[datetime] = None
    endTime: Optional[datetime] = None
    photoCount: int = 0
    coverPhotoId: Optional[str] = None
    coverAssetId: Optional[str] = None
    coverShootTime: Optional[datetime] = None
    coverGpsLat: Optional[float] = None
    coverGpsLon: Optional[float] = None
    coverPhotoUrl: Optional[str] = None
    storyText: Optional[str] = None
    fullStory: Optional[str] = None
    detailedLocation: Optional[str] = None
    locationTags: Optional[str] = None
    emotionTag: Optional[str] = None
    musicUrl: Optional[str] = None
    status: EventStatus = "clustered"
    eventVersion: int = 1
    storyGeneratedFromVersion: Optional[int] = None
    storyFreshness: StoryFreshness = "stale"
    slideshowGeneratedFromVersion: Optional[int] = None
    slideshowFreshness: StoryFreshness = "stale"
    hasPendingStructureChanges: bool = True
    titleManuallySet: bool = False
    storyReady: bool = False
    visionSummary: EventVisionSummary = Field(default_factory=EventVisionSummary)
    aiError: Optional[str] = None
    updatedAt: Optional[datetime] = None


class EventEnhancementSummary(BaseModel):
    status: EnhancementStatus = "none"
    assetCount: int = 0
    totalBytes: int = 0
    canRetry: bool = False
    lastUploadedAt: Optional[datetime] = None
    retainedUntil: Optional[datetime] = None


class EnhancementStorageSummary(BaseModel):
    eventCount: int = 0
    assetCount: int = 0
    totalBytes: int = 0
    nextExpiresAt: Optional[datetime] = None


class EventListResponse(BaseModel):
    items: list[EventResponse]
    total: int
    page: int
    pageSize: int
    totalPages: int


class EventPhotoItem(BaseModel):
    id: str
    assetId: Optional[str] = None
    fileHash: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None
    photoUrl: Optional[str] = None
    thumbnailUrl: Optional[str] = None
    shootTime: Optional[datetime] = None
    gpsLat: Optional[float] = None
    gpsLon: Optional[float] = None
    caption: Optional[str] = None
    photoIndex: Optional[int] = None
    visualDesc: Optional[str] = None
    microStory: Optional[str] = None
    emotionTag: Optional[str] = None
    visionStatus: str = "pending"
    visionError: Optional[str] = None
    visionUpdatedAt: Optional[datetime] = None
    vision: Optional[PhotoVisionResult] = None


class EventDetailResponse(EventResponse):
    photos: list[EventPhotoItem] = Field(default_factory=list)
    chapters: list[EventChapterResponse] = Field(default_factory=list)
    photoGroups: list[PhotoGroupResponse] = Field(default_factory=list)
    enhancement: EventEnhancementSummary = Field(default_factory=EventEnhancementSummary)


class EventCreateRequest(BaseModel):
    title: Optional[str] = None
    locationName: Optional[str] = None
    photoIds: list[str] = Field(default_factory=list)


class EventUpdateRequest(BaseModel):
    title: Optional[str] = None
    locationName: Optional[str] = None
    gpsLat: Optional[float] = None
    gpsLon: Optional[float] = None
    coverPhotoUrl: Optional[str] = None
    storyText: Optional[str] = None
    fullStory: Optional[str] = None
    detailedLocation: Optional[str] = None
    locationTags: Optional[str] = None
    emotionTag: Optional[str] = None
    musicUrl: Optional[str] = None
    status: Optional[EventStatus] = None


class RegenerateStoryResponse(BaseModel):
    taskId: Optional[str] = None
    status: str


class EnhanceStoryResponse(BaseModel):
    taskId: Optional[str] = None
    status: str
    enhancement: EventEnhancementSummary


class LocationCityCandidate(BaseModel):
    name: str
    displayName: str
    adcode: str


class LocationPlaceCandidate(BaseModel):
    name: str
    address: str = ""
    locationName: str
    detailedLocation: str
    locationTags: str = ""
    gpsLat: float
    gpsLon: float

File: mobile/src/types/event.ts
import type { EventChapter } from '@/types/chapter';
import type { PhotoGroup } from '@/types/photoGroup';
import type { OnDeviceVisionResult } from '@/types/vision';

export type EventStatus =
  | 'clustered'
  | 'waiting_for_vision'
  | 'ai_pending'
  | 'ai_processing'
  | 'generated'
  | 'ai_failed';
export type EventEnhancementStatus = 'none' | 'retained' | 'expired';
export type StoryFreshness = 'fresh' | 'stale';
export type EventVisionStatus =
  | 'pending'
  | 'processing'
  | 'partial'
  | 'completed'
  | 'failed'
  | 'unsupported';

export interface EventVisionSummary {
  status: EventVisionStatus;
  total: number;
  pending: number;
  processing: number;
  completed: number;
  failed: number;
  unsupported: number;
}

export interface EventEnhancementSummary {
  status: EventEnhancementStatus;
  assetCount: number;
  totalBytes: number;
  canRetry: boolean;
  lastUploadedAt?: string | null;
  retainedUntil?: string | null;
}

export interface EnhancementStorageSummary {
  eventCount: number;
  assetCount: number;
  totalBytes: number;
  nextExpiresAt?: string | null;
}

export interface EventRecord {
  id: string;
  title: string;
  locationName?: string | null;
  gpsLat?: number | null;
  gpsLon?: number | null;
  startTime?: string | null;
  endTime?: string | null;
  photoCount: number;
  coverPhotoId?: string | null;
  coverAssetId?: string | null;
  coverShootTime?: string | null;
  coverGpsLat?: number | null;
  coverGpsLon?: number | null;
  localCoverUri?: string | null;
  selectedCoverPhotoId?: string | null;
  coverPhotoUrl?: string | null;
  storyText?: string | null;
  fullStory?: string | null;
  heroTitle?: string | null;
  heroSummary?: string | null;
  detailedLocation?: string | null;
  locationTags?: string | null;
  emotionTag?: string | null;
  musicUrl?: string | null;
  status: EventStatus;
  eventVersion: number;
  storyGeneratedFromVersion?: number | null;
  storyFreshness: StoryFreshness;
  slideshowGeneratedFromVersion?: number | null;
  slideshowFreshness: StoryFreshness;
  hasPendingStructureChanges: boolean;
  titleManuallySet: boolean;
  storyReady: boolean;
  visionSummary: EventVisionSummary;
  aiError?: string | null;
  updatedAt?: string | null;
}

export interface EventListResult {
  items: EventRecord[];
  total: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

export interface EventPhotoItem {
  id: string;
  fileHash?: string | null;
  assetId?: string | null;
  width?: number | null;
  height?: number | null;
  localUri?: string | null;
  localThumbnailUri?: string | null;
  localCoverUri?: string | null;
  photoUrl?: string | null;
  thumbnailUrl?: string | null;
  shootTime?: string | null;
  gpsLat?: number | null;
  gpsLon?: number | null;
  storyText?: string | null;
  caption?: string | null;
  photoIndex?: number | null;
  visualDesc?: string | null;
  microStory?: string | null;
  emotionTag?: string | null;
  visionStatus?: 'pending' | 'processing' | 'completed' | 'failed' | 'unsupported' | null;
  visionError?: string | null;
  visionUpdatedAt?: string | null;
  vision?: OnDeviceVisionResult | null;
}

export interface EventDetail extends EventRecord {
  photos: EventPhotoItem[];
  chapters: EventChapter[];
  photoGroups: PhotoGroup[];
  enhancement?: EventEnhancementSummary | null;
}

export interface RegenerateStoryResult {
  taskId?: string | null;
  status: 'queued' | 'processed_inline';
}

export interface EnhanceStoryResult {
  taskId?: string | null;
  status: 'queued' | 'processed_inline';
  enhancement: EventEnhancementSummary;
}
